about_arrays.rb

```diff
@@ -0,0 +1,84 @@
+require File.expand_path(File.dirname(__FILE__) + '/neo')
+
+class AboutArrays < Neo::Koan
+  def test_creating_arrays
+    empty_array = Array.new
+    assert_equal Array, empty_array.class
+    assert_equal 0, empty_array.size
+  end
+
+  def test_array_literals
+    array = Array.new
+    assert_equal [], array
+
+    array[0] = 1
+    assert_equal [1], array
+
+    array[1] = 2
+    assert_equal [1, 2], array
+
+    array << 333
+    assert_equal [1, 2, 333], array
+  end
+
+  def test_accessing_array_elements
+    array = [:peanut, :butter, :and, :jelly]
+
+    assert_equal :peanut, array[0]
+    assert_equal :peanut, array.first
+    assert_equal :jelly, array[3]
+    assert_equal :jelly, array.last
+    assert_equal :jelly, array[-1]
+    assert_equal :butter, array[-3]
+  end
+
+  def test_slicing_arrays
+    array = [:peanut, :butter, :and, :jelly]
+
+    assert_equal [:peanut], array[0,1]
+    assert_equal [:peanut, :butter], array[0,2]
+    assert_equal [:and, :jelly], array[2,2]
+    assert_equal [:and, :jelly], array[2,20]
+    assert_equal [], array[4,0]
```

Comment:
@stevendnoble here is some **very** irrelevant information but it may help guide seeing where these language decisions are made.

So the array behavior is somewhat encapsulated at this line (in Ruby 2):

https://github.com/ruby/ruby/blob/65c744fb9cb65222dbab16497444da87df86966f/array.c#L1270

But it starts getting interesting as the use case you're using is part of this function:

https://github.com/ruby/ruby/blob/65c744fb9cb65222dbab16497444da87df86966f/array.c#L1205

The logic checks if the beginning is greater than the length of the array:

```c
    if (beg > alen) return Qnil;
```

But it doesn't have a case to see if the `beg == alen` which means it falls to the last scenario which is to create an empty array:

```c
if (len == 0) return ary_new(klass, 0);
```

So I'd say that maybe when they were putting this in it was a mistake which became a language feature.